File: src/whoami/collectors/codex.py
from __future__ import annotations

from datetime import date, datetime
from pathlib import Path

from whoami.collectors.jsonl_util import clip, iter_jsonl, on_day, parse_ts
from whoami.models import SessionEvent, TokenUsage
from whoami.paths import home


def collect(day: date) -> list[SessionEvent]:
    roots = [home() / ".codex" / "sessions", home() / ".codex" / "archived_sessions"]
    sessions: list[SessionEvent] = []
    for root in roots:
        if not root.exists():
            continue
        for path in root.rglob("*.jsonl"):
            event = _parse_file(path, day)
            if event:
                sessions.append(event)
    return sessions


def _parse_file(path: Path, day: date) -> SessionEvent | None:
    tokens = TokenUsage()
    prompts: list[str] = []
    files: set[str] = set()
    tools: set[str] = set()
    times: list[datetime] = []
    model: str | None = None
    session_id = path.stem
    project = path.parent.name

    relevant = False
    for obj in iter_jsonl(path):
        payload = obj.get("payload") if isinstance(obj.get("payload"), dict) else obj
        ts = parse_ts(
            obj.get("timestamp")
            or payload.get("timestamp")
            or obj.get("created_at")
            or payload.get("created_at")
        )
        if ts and on_day(ts, day):
            relevant = True
            times.append(ts)

        kind = obj.get("type") or payload.get("type") or obj.get("record_type")
        model = payload.get("model") or obj.get("model") or model

        usage = payload.get("token_usage") or payload.get("usage") or obj.get("usage")
        if isinstance(usage, dict):
            tokens = tokens.add(_usage(usage))

        text = ""
        if kind in {"user_message", "user", "event_msg"} or payload.get("role") == "user":
            text = _as_text(payload.get("text") or payload.get("content") or obj.get("text"))
        if text and ts and on_day(ts, day):
            prompts.append(clip(text))

        tool_name = payload.get("tool") or payload.get("name")
        if isinstance(tool_name, str) and tool_name:
            tools.add(tool_name)
        for key in ("path", "file_path", "cwd"):
            val = payload.get(key)
            if isinstance(val, str) and ("/" in val or "\\" in val):
                files.add(val)

    if not relevant:
        return None

    return SessionEvent(
        tool="codex",
        session_id=session_id,
        project=project,
        started_at=min(times) if times else None,
        ended_at=max(times) if times else None,
        model=model,
        tokens=tokens,
        user_prompts=prompts[:12],
        files_touched=sorted(files)[:30],
        tools_used=sorted(tools),
        title=prompts[0] if prompts else session_id,
    )


def _usage(usage: dict) -> TokenUsage:
    return TokenUsage(
        input_tokens=int(
            usage.get("input_tokens") or usage.get("input") or usage.get("prompt_tokens") or 0
        ),
        output_tokens=int(
            usage.get("output_tokens") or usage.get("output") or usage.get("completion_tokens") or 0
        ),
        cache_read_tokens=int(
            usage.get("cached_input_tokens")
            or usage.get("cache_read_tokens")
            or usage.get("cache_read")
            or 0
        ),
        cache_write_tokens=int(usage.get("cache_write_tokens") or usage.get("cache_write") or 0),
    )


def _as_text(content: object) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return " ".join(_as_text(x) for x in content if x)
    if isinstance(content, dict):
        return str(content.get("text") or content.get("content") or "")
    return ""

File: packages/core/src/collectors/codex.ts
import { existsSync } from "node:fs";
import { basename, join } from "node:path";
import { home } from "../paths";
import { iterJsonl, rec, walkFiles } from "../fsutil";
import { asText, clip, displayProject, intish, pickTitle } from "../text";
import { clusterTimestamps, onDay, parseTs, toIso } from "../time";
import { addTokens, emptyTokens, type SessionEvent, type TokenUsage } from "../types";

function codexRoots(): string[] {
  const homeCodex = join(home(), ".codex");
  return [join(homeCodex, "sessions"), join(homeCodex, "archived_sessions")];
}

export async function collectCodex(day: string, timeZone: string): Promise<SessionEvent[]> {
  const sessions: SessionEvent[] = [];
  for (const root of codexRoots()) {
    if (!existsSync(root)) continue;
    const files = await walkFiles(root, (name) => name.endsWith(".jsonl"));
    for (const path of files) {
      const event = await parseFile(path, day, timeZone);
      if (event) sessions.push(event);
    }
  }
  return sessions;
}

function usageFrom(raw: Record<string, unknown> | null): TokenUsage {
  if (!raw) return emptyTokens();
  return {
    input_tokens: intish(raw.input_tokens ?? raw.input ?? raw.prompt_tokens),
    output_tokens: intish(raw.output_tokens ?? raw.output ?? raw.completion_tokens),
    cache_read_tokens: intish(raw.cached_input_tokens ?? raw.cache_read_tokens ?? raw.cache_read),
    cache_write_tokens: intish(raw.cache_write_tokens ?? raw.cache_write),
  };
}

function isUserTurn(kind: string, payload: Record<string, unknown>): boolean {
  if (payload.role === "user") return true;
  if (kind === "user_message" || kind === "user") return true;
  if (payload.type === "user_message") return true;
  if (kind === "event_msg" || payload.role === "developer" || payload.role === "system") return false;
  return false;
}

async function parseFile(path: string, day: string, timeZone: string): Promise<SessionEvent | null> {
  let tokens = emptyTokens();
  const prompts: string[] = [];
  const files = new Set<string>();
  const tools = new Set<string>();
  const times: Date[] = [];
  let model: string | null = null;
  let cwd: string | null = null;
  let sessionId = basename(path).replace(/\.jsonl$/, "");

  for await (const obj of iterJsonl(path)) {
    const payload = rec(obj.payload) || obj;
    const ts = parseTs(obj.timestamp ?? payload.timestamp ?? obj.created_at ?? payload.created_at);
    const kind = String(obj.type ?? payload.type ?? obj.record_type ?? "");
    if (kind === "session_meta") {
      if (typeof payload.session_id === "string") sessionId = payload.session_id;
      if (typeof payload.cwd === "string") cwd = payload.cwd;
      if (typeof payload.model === "string") model = payload.model;
    }
    if (typeof payload.model === "string") model = payload.model;

    if (ts && onDay(ts, day, timeZone)) {
      times.push(ts);
      const usage = rec(payload.token_usage) || rec(payload.usage) || rec(obj.usage);
      if (usage) tokens = addTokens(tokens, usageFrom(usage));
      if (isUserTurn(kind, payload)) {
        const text = asText(payload.text ?? payload.content ?? obj.text);
        if (text) prompts.push(clip(text));
      }
      const toolName = payload.tool ?? payload.name;
      if (typeof toolName === "string" && toolName) tools.add(toolName);
      for (const key of ["path", "file_path"]) {
        const val = payload[key];
        if (typeof val === "string" && (val.includes("/") || val.includes("\\"))) files.add(val);
      }
    }
  }

  if (!times.length) return null;
  const clusters = clusterTimestamps(times);
  const hours = clusters.reduce((s, c) => s + c.hours, 0);
  return {
    tool: "codex",
    session_id: sessionId,
    project: displayProject(cwd) || displayProject(basename(path)),
    started_at: toIso(clusters[0]?.start ?? null),
    ended_at: toIso(clusters.at(-1)?.end ?? null),
    model,
    tokens,
    user_prompts: prompts.slice(0, 12),
    files_touched: [...files].slice(0, 30),
    tools_used: [...tools].sort(),
    title: pickTitle(prompts, sessionId),
    hours,
  };
}
